jib-core/src/main/java/com/google/cloud/tools/jib/ncache/DefaultCacheStorageWriter.java

```diff
@@ -0,0 +1,236 @@
+   *
+   * @param cacheWrite the {@link CacheWrite} to write out
+   * @return the {@link CacheEntry} representing the written entry
+   * @throws IOException if an I/O exception occurs
+   */
+  CacheEntry write(CacheWrite cacheWrite) throws IOException {
+    // Creates the layers directory if it doesn't exist.
+    Files.createDirectories(defaultCacheStorageFiles.getLayersDirectory());
+
+    // Creates the temporary directory.
+    try (TemporaryDirectory temporaryDirectory = new TemporaryDirectory()) {
+      Path temporaryLayerDirectory = temporaryDirectory.getDirectory();
+
+      // Writes the layer file to the temporary directory.
+      WrittenLayer writtenLayer =
+          writeLayerBlobToDirectory(cacheWrite.getLayerBlob(), temporaryLayerDirectory);
+
+      // Writes the metadata to the temporary directory.
+      if (cacheWrite.getMetadataBlob().isPresent()) {
+        writeMetadataBlobToDirectory(cacheWrite.getMetadataBlob().get(), temporaryLayerDirectory);
+      }
+
+      // Moves the temporary directory to the final location.
+      moveIfDoesNotExist(
+          temporaryLayerDirectory,
+          defaultCacheStorageFiles.getLayerDirectory(writtenLayer.layerDigest));
+
+      // Updates cacheEntry with the blob information.
+      Path layerFile =
+          defaultCacheStorageFiles.getLayerFile(writtenLayer.layerDigest, writtenLayer.layerDiffId);
+      DefaultCacheEntry.Builder cacheEntryBuilder =
+          DefaultCacheEntry.builder()
+              .setLayerDigest(writtenLayer.layerDigest)
+              .setLayerDiffId(writtenLayer.layerDiffId)
+              .setLayerSize(writtenLayer.layerSize)
+              .setLayerBlob(Blobs.from(layerFile));
+      if (cacheWrite.getMetadataBlob().isPresent()) {
+        Path metadataFile = defaultCacheStorageFiles.getMetadataFile(writtenLayer.layerDigest);
+        cacheEntryBuilder.setMetadataBlob(Blobs.from(metadataFile)).build();
+      }
+
+      // Write the selector file.
+      if (cacheWrite.getSelector().isPresent()) {
+        writeSelector(cacheWrite.getSelector().get(), writtenLayer.layerDigest);
+      }
+
+      return cacheEntryBuilder.build();
+    }
+  }
+
+  /**
+   * Writes the {@code layerBlob} to the {@code layerDirectory}.
+   *
+   * @param layerBlob the layer {@link Blob}
+   * @param layerDirectory the directory for the layer
+   * @return a {@link WrittenLayer} with the written layer information
+   * @throws IOException if an I/O exception occurs
+   */
+  private WrittenLayer writeLayerBlobToDirectory(Blob layerBlob, Path layerDirectory)
+      throws IOException {
+    Path temporaryLayerFile = Files.createTempFile(layerDirectory, null, null);
+    temporaryLayerFile.toFile().deleteOnExit();
+
+    try (CountingDigestOutputStream compressedDigestOutputStream =
+        new CountingDigestOutputStream(
+            new BufferedOutputStream(FileOperations.newLockingOutputStream(temporaryLayerFile)))) {
```

Comment:
Sorry I missed this, but it doesn't seem necessary to use locking outputstream on temporary files. Shouldn't their uniqueness be provided by guava?

Reply:
Oh yes, that's true. Will change this in #898